### PR TITLE
Add timeout for ReadyToTest action

### PR DESCRIPTION
**Pull request by deepanshubansal01**
Related to #582 

Signed-off-by: deepanshu <deepanshubansal01@gmail.com>

### CODE REVIEW

**Review by adityapande-1995 — COMMENTED** [review]
Can we have a test that uses multiple runners or Launch descriptions ? Probably a parameterized test here ? https://github.com/ros2/launch/blob/69964b72cf586be99b0b1d9fcfe8b68545123fff/launch_testing/test/launch_testing/test_launch_test_runner_validation.py#L157

File: launch_testing/test/launch_testing/test_launch_test_runner_validation.py [hunk 1]
```diff
@@ -167,3 +167,24 @@ def generate_test_description(my_param):
...
+        runs = make_test_run_for_dut(generate_test_description)
+        dut = LaunchTestRunner(
+            test_runs=runs,
+            launch_file_arguments=[],
+            debug=False,
+            timeout=20.0  # increase the timeout giving the time for process to startup
+        )
+
+        dut.validate()  # Make sure this passes initial validation (probably redundant with above)
+        runs[0].normalized_test_description(ready_fn=lambda: None)
```
Comment (adityapande-1995) [inline launch_testing/test/launch_testing/test_launch_test_runner_validation.py:190]: Flake8 needs the newline character here it seems.

**Comment by deepanshubansal01** [comment]
> Reference

I can try that but is this test for checking if we can we have multiple runners/launch descriptions or for testing the timeout thing? 
Also for the params I am not sure if they are the same as cli optional arguments parsed using argsparse lib?

File: launch_testing/test/launch_testing/test_launch_test_runner_validation.py [hunk 2]
```diff
@@ -153,7 +146,6 @@ def generate_test_description():
         runs[0].normalized_test_description(ready_fn=lambda: None)
 
     def test_parametrized_launch_description(self):
-
```
[the remark this answers is not in the document]
Comment (deepanshubansal01) [reply launch_testing/test/launch_testing/test_launch_test_runner_validation.py]: Done

File: launch_testing/launch_testing/launch_test.py [hunk 3]
```diff
@@ -56,7 +56,10 @@ def add_arguments(parser):
         '--junit-xml', action='store', dest='xmlpath', default=None,
         help='Do write xUnit reports to specified path.'
     )
-
+    parser.add_argument(
+        '--timeout', type=float, default=15.0,
+        help='timeout for ReadyToTest action'
```
Comment (jacobperron) [inline launch_testing/launch_testing/launch_test.py:61]: I think we should clarify a few things in the help message:
- this is the duration between the start of the launch process and the ReadyToTest action being executed.
- what happens if the timeout occurs
- the units are seconds

File: launch_testing/test/launch_testing/test_launch_test_runner_validation.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def make_test_run_for_dut(generate_test_description_function):
 class TestLaunchTestRunnerValidation(unittest.TestCase):
 
     def test_catches_bad_signature(self):
-
+
```
Comment (jacobperron) [inline launch_testing/test/launch_testing/test_launch_test_runner_validation.py:36]: I guess linters won't like the extra whitespace introduced here.

File: launch_testing/test/launch_testing/test_launch_test_runner_validation.py [hunk 5]
```diff
@@ -167,3 +167,23 @@ def generate_test_description(my_param):
 
         dut.validate()  # Make sure this passes initial validation (probably redundant with above)
         runs[0].normalized_test_description(ready_fn=lambda: None)
+
+    def test_launch_description_with_ready_action_changed_timeout(self):
```
Comment (jacobperron) [inline launch_testing/test/launch_testing/test_launch_test_runner_validation.py:171]: I don't think this test is actually testing the timeout. It only performs validation of the launch description, but does not actually execute anything as far as I can tell.

I see a few other tests that exercise some of the other command line options, maybe we can do something similar for the `--timeout` option:

- https://github.com/ros2/launch/blob/master/launch_testing/test/launch_testing/test_print_arguments.py
- https://github.com/ros2/launch/blob/master/launch_testing/test/launch_testing/test_xml_output.py

**Review by jacobperron — COMMENTED** [review]
The new option looks good to me :+1:

File: launch_testing/test/launch_testing/test_launch_test_runner_validation.py (hunk 5 above)
Comment (deepanshubansal01) [reply launch_testing/test/launch_testing/test_launch_test_runner_validation.py:171]: I see.

**Review by ivanpauno — CHANGES_REQUESTED** [review]
IMO, it would be nice if the timeout is somehow specified in the launch file, instead of being an option for the as a command line argument.

Is there any alternative to make that happen?

**Comment by deepanshubansal01** [comment]
> IMO, it would be nice if the timeout is somehow specified in the launch file, instead of being an option for the as a command line argument.
> 
> Is there any alternative to make that happen?

I tried making timeout configurable as part of launch file itself and having timeout as an argument to the `ReadyToTest` action. However, this is not possible with the current code base. This is because the `_run_test` thread which checks for `ReadyToTest` Action timeout is started in the `_RunnerWorker` class constructor and the `ReadyToTest` action itself is instantiated after the `_run_test` thread has been started which makes it futile to access the timeout attribute from `ReadyToTest` action class since the `_run_test` thread has already been started. 

The `_run_test` thread is started here: https://github.com/ros2/launch/blob/074cd2903ddccd61bce8f40a0f58da0b7c200481/launch_testing/launch_testing/test_runner.py#L60-L64
And the `ReadyToTest` object is instantiated here: https://github.com/ros2/launch/blob/074cd2903ddccd61bce8f40a0f58da0b7c200481/launch_testing/launch_testing/test_runner.py#L73-L75

**Comment by deepanshubansal01** [comment]
> > IMO, it would be nice if the timeout is somehow specified in the launch file, instead of being an option for the as a command line argument.
> 
> How would the timeout parameter being added here being specified in practice? i.e. where would the users specify it? (users are not usually running launch_testing manually)

I am not sure about this right now the timeout can be passed through cli and manually to the `LaunchTestRunner` class. If the use case is not through cli then we might have to do something else? Passing it in launch file itself makes much more sense I guess as part of `ReadyToTest` action but I am not sure how can we do it with the current design.

**Comment by ivanpauno** [comment]
Ideally, I think that a decorator would be a nice way to specify it:

```python3
@launch_testing.runner_processes_wait_timeout(5.3)
def generate_test_description(arg_param):
```

To have something like that, the decorator would have to add a hidden attribute to the decorated function (e.g. `generate_test_description.__runner_processes_wait_timeout`).
The loader could read that attribute, similar to how [parametrized](https://github.com/ros2/launch/blob/dcf180a0da309a1c24c210f56d381fc59ea1546a/launch_testing/launch_testing/loader.py#L164) launch descriptions are handled.
Each [TestRun](https://github.com/ros2/launch/blob/dcf180a0da309a1c24c210f56d381fc59ea1546a/launch_testing/launch_testing/loader.py#L77) could have the timeout as a property.
And then the runner would use `self._test_run.runner_processes_wait_timeout` instead of `15` [here](https://github.com/ros2/launch/blob/dcf180a0da309a1c24c210f56d381fc59ea1546a/launch_testing/launch_testing/test_runner.py#L184).

@deepanshubansal01 @jacobperron does that make sense to you?

**Comment by ivanpauno** [comment]
@hidmic maybe you can provide more feedback of how to implement this

**Comment by deepanshubansal01** [comment]
> @hidmic maybe you can provide more feedback of how to implement this

@ivanpauno should I proceed with the implementation that you suggested or wait for @hidmic inputs. Thanks

**Comment by ivanpauno** [comment]
> @ivanpauno should I proceed with the implementation that you suggested or wait for @hidmic inputs.

I think it's fine to proceed with the implementation.

**Comment by deepanshubansal01** [comment]
Closing the PR now, the feature will be implemented in this PR now https://github.com/ros2/launch/pull/625

**Pull request closed without merging**